program-libs/zero-copy/src/vec.rs

```
@@ -0,0 +1,326 @@
+    pub fn data_size(length: LEN) -> usize {
+        ZeroCopySliceMut::<LEN, T>::required_size_for_capacity(length.try_into().unwrap())
+    }
+
+    #[inline]
+    pub fn required_size_for_capacity(capacity: usize) -> usize {
+        Self::metadata_size() + Self::data_size(capacity.try_into().unwrap())
+    }
+}
+
+impl<LEN, T> ZeroCopyVec<LEN, T>
+where
+    LEN: TryFrom<usize> + TryInto<usize> + Copy + Clone + Add<LEN, Output = LEN>,
+    T: Copy + Clone,
+    <LEN as TryFrom<usize>>::Error: fmt::Debug,
+    <LEN as TryInto<usize>>::Error: fmt::Debug,
+{
+    #[inline]
+    pub fn len(&self) -> usize {
+        (*self.length).try_into().unwrap()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[inline]
+    pub fn get(&self, index: usize) -> Option<&T> {
+        if index >= self.len() {
+            return None;
+        }
+        Some(&self.data[index])
+    }
+
+    #[inline]
+    pub fn get_mut(&mut self, index: usize) -> Option<&mut T> {
+        if index >= self.len() {
+            return None;
+        }
+        Some(&mut self.data[index])
+    }
+
+    #[inline]
+    pub fn first(&self) -> Option<&T> {
+        self.get(0)
+    }
+
+    #[inline]
+    pub fn first_mut(&mut self) -> Option<&mut T> {
+        self.get_mut(0)
+    }
+
+    #[inline]
+    pub fn last(&self) -> Option<&T> {
+        self.get(self.len().saturating_sub(1))
+    }
+
+    #[inline]
+    pub fn last_mut(&mut self) -> Option<&mut T> {
+        self.get_mut(self.len().saturating_sub(1))
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[T] {
+        &self.data.as_slice()[..self.len()]
+    }
+
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        let len = self.len();
+        &mut self.data.as_mut_slice()[..len]
+    }
+
+    pub fn copy_from_slice(&mut self, slice: &[T]) {
+        let len = self.len();
+        let new_len = len + slice.len();
+        if new_len > self.capacity() {
```

It's confusing, shouldn't it be instead `if slice.len() != self.len() {`?

---

or, instead, on line 208: `self.data.as_mut_slice()[len..new_len].copy_from_slice(slice);`, but in this case method should be renamed to `extend_from_slice`

---

you are right copy_from_slice should only work with exact length.
std slice is implemented like this as well:
https://doc.rust-lang.org/src/core/slice/mod.rs.html#3691-3693

---

> if slice.len() != self.len() 
is probably better so that we can throw an explicit error